Search Filter Split, Use Same Controller

askvortsov1 opened the pull request:
**Start of flarum/issue-archive#286**
Supercedes flarum/framework#2453 

**Changes proposed in this pull request:**
- Split search and filter into separate logic
- Add a "filterer" that applies filters and filter mutations for all resources
- Add filter equivalent of default search gambits
- Add integration tests for all filters and gambits present in core (this is probably 75% of the PR)

TODO:
- Convert bundled extensions to use this system
- Clean up code / improve abstractions
- Convert ListPostsController to use this system (in a separate PR)

**Reviewers should focus on:**
- If we register things both as filters and as gambits for SimpleFlarumSearch, we're gonna have a lot of duplicated code. That being said, I'm not sure if there's anything we can do here if we want to truly separate the two. Maybe gambits could implement both interfaces?
- For simplicity and to reduce diff size, I omitted some of the abstractions present in the old search system for filtering (SearchCriteria, etc). I also reused some SimpleFlarumSearch utils / components, like SearchResults, ApplySearchParametersTrait, AbstractSearch. What do we want to do about this? IMO simpler is better, especially seeing that custom filtration drivers aren't going to be a thing, so unnecessary abstraction is unnecessary. Did I cut out too much? Also, I want to eventually move the shared stuff into a Query namespace, but didn't do so here to minimize diff.
- Should we modify the new gambits system to also be applied based on key/value pairs in the filter param? And q could just be the fulltext search?

askvortsov1 committed:
Introduce extensible filterer system, make ListUsersController and ListDiscussionsController use that instead of searcher if a "q" filter parameter is provided

askvortsov1 committed:
Add filter counterparts to builtin gambits

askvortsov1 committed:
Properly register filters, support negation (passed into filter interface)

askvortsov1 committed:
Add tests for all user filters and gambits

askvortsov1 committed:
Fix created filter

askvortsov1 committed:
Add tests for all discussion filters and gambits

askvortsov1 committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

luceos reviewed (changes requested):
The arguments for the controller construct should be similar. In the DiscussionController the repository is the first argument, in the UsersController it's the last.

Great job!

Aside from that I see nothing wrong with this. How this impacts extensions is still a bit unclear and I think seeing it in action will give more insight in whether this is the best approach or not.

askvortsov1 committed:
Fix failing tests

askvortsov1 committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

SychO9 reviewed (changes requested):
I don't see any issues with using a single endpoint, I think it'll work just fine.

I'm leaning towards having a single class implementing both the gambit and the filter interfaces, even if we don't reduce code duplication, having them in one place would be better I think.

> That would make the API for filtering much more similar to the API for searching, but isn't strictly necessary. Any thoughts

I would be in favour. It's more consistent with the search API, but also cleaner imho. Although having a filterer for each resource would mean more duplicated code similarly to resource searchers 🤔

askvortsov1 committed:
Add types

askvortsov1 committed:
Expect eloquent builder

askvortsov1 committed:
Merge gambits into "FilterGambit"

askvortsov1 committed:
Use Arr::get in CreatedFilterGambit

askvortsov1 committed:
Inject filters and filter mutators instead of static

askvortsov1 committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

askvortsov1 committed:
Per-resource filter classes

askvortsov1 committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

askvortsov1 committed:
List controllers no longer need repositories

askvortsov1 committed:
Add missing imports

askvortsov1 committed:
Move gambit pattern to method instead of attribute for consistency

askvortsov1 committed:
Combine search and filter criteria classes

askvortsov1 committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

askvortsov1 committed:
Use null coalesce instead of shorthand ternary

askvortsov1 committed:
Replace WrappedFilter and AbstractSearch with FilterState and SearchState

askvortsov1 committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

askvortsov1 committed:
Remove unnecessary load

This is already done in the searcher

askvortsov1 committed:
Rename $load to $include

$load is the mechanism, $include is the actual important bit.

askvortsov1 committed:
Typehint AbstractSearcher return value

askvortsov1 committed:
Move $include out of searching / filtering

On second thought,  the searcher / filterer shouldn't be responsible for loading related items. It centralizes code, but isn't proper (and breaks `Discussion::setStateUser()`)

askvortsov1 committed:
Fix stupid mistakes

askvortsov1 committed:
More fixes for stupid stuff, also temporarily loosen interface to avoid BC breaks

askvortsov1 committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

askvortsov1 committed:
Don't pass unused $include to filterer/searcher

askvortsov1 committed:
Remove includes param from AbstractSearcher

askvortsov1 committed:
Add filter mutator params to docblock

askvortsov1 committed:
Capitalization fix

SychO9 reviewed (approved):
This must be the biggest change in this release. We might not want to merge it right away when we get the two approvals, in case any of the other devs are able to review it in time.

tankerkiller125 reviewed (approved):
Have not tested locally, but it looks good code wise as far as I can tell
